test/test_jit_fuser_te.py

```diff
@@ -1827,6 +1827,25 @@ def repro(xs: List[torch.Tensor], ys: List[torch.Tensor], zs: List[torch.Tensor]
                 zs = [torch.ones(i) for i in range(N)]
                 repro(xs, ys, zs)
 
+    def test_scalar_and_tensor_constant(self):
+        @torch.jit.script
+        def foo(a, c):
+            b = 0.0
+            if bool(a == 0.0):
+                b = 1.0
+            return b + c
+
+        a = torch.ones(1, dtype=torch.float)
+
+        def use(b):
+            return foo(b - 1.0, a) + 1.0
+        use = torch.jit.trace(use, (torch.zeros(1, dtype=torch.float),))
+
+        # test we propagated shapes through the function
+        self.assertTrue("Dynamic" not in str(use.graph))
```

Comment:
Dynamic shapes are not printed as "Dynamic", they are printed as "Tensor".
In general, I'm a bit suspicious about this test - do you mind elaborating what we're trying to check here? Do we really need two functions to check how scalars are handled? Can we simply parse a graph with scalar inputs from text? Another reason why I'm suspicious is that AFAICT we shouldn't be fusing scalars at all until #59279 lands.

Reply:
I lifted this test verbatim from `test_tracer.py test_trace_of_script`, so that we'd get OSS coverage of the fuser in that case.  But I supposed I can write a more minimal test case here.

I guess I usually prefer writing python code for tests rather than using the IR parser, but I can see the advantages.

I don't think we've ever prevented scalar fusion have we?  What's new here is that the *only* input to the graph is a scalar, because the tensor operations are performed on a constant.

Reply:
> I don't think we've ever prevented scalar fusion have we?

We did:
https://github.com/pytorch/pytorch/blob/2119efd234ff80b2a86b5cd9d15fb29d1e18de49/torch/csrc/jit/passes/tensorexpr_fuser.cpp#L850-L852